File: core/services/gcs/src/copier.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::sync::Arc;

use bytes::Buf;

use super::core::GcsCore;
use super::core::RewriteResponse;
use super::core::constants::GCS_REWRITE_MAX_CHUNK_SIZE;
use super::core::constants::GCS_REWRITE_MIN_CHUNK_SIZE;
use super::error::parse_error;
use opendal_core::raw::*;
use opendal_core::*;

pub struct GcsCopier {
    core: Arc<GcsCore>,
    ctx: OperationContext,
    from: String,
    to: String,
    args: OpCopy,

    chunk: Option<usize>,
    rewrite_token: Option<String>,
    total_bytes_rewritten: u64,
    completed: bool,
    metadata: Option<Metadata>,
}

impl GcsCopier {
    pub fn new(
        core: Arc<GcsCore>,
        ctx: OperationContext,
        from: &str,
        to: &str,
        args: OpCopy,
        opts: OpCopier,
    ) -> Self {
        let chunk = opts.chunk().map(|v| {
            let v = v.clamp(GCS_REWRITE_MIN_CHUNK_SIZE, GCS_REWRITE_MAX_CHUNK_SIZE);
            v / GCS_REWRITE_MIN_CHUNK_SIZE * GCS_REWRITE_MIN_CHUNK_SIZE
        });

        Self {
            core,
            ctx,
            from: from.to_string(),
            to: to.to_string(),
            args,
            chunk,
            rewrite_token: None,
            total_bytes_rewritten: 0,
            completed: false,
            metadata: None,
        }
    }
}

impl oio::Copy for GcsCopier {
    async fn next(&mut self) -> Result<Option<usize>> {
        if self.completed {
            return Ok(None);
        }

        let resp = self
            .core
            .gcs_rewrite_object(
                &self.ctx,
                &self.from,
                &self.to,
                &self.args,
                self.chunk,
                self.rewrite_token.as_deref(),
            )
            .await?;

        if !resp.status().is_success() {
            return Err(parse_error(resp));
        }

        let result: RewriteResponse = serde_json::from_reader(resp.into_body().reader())
            .map_err(new_json_deserialize_error)?;

        let total = result.total_bytes_rewritten.parse::<u64>().map_err(|err| {
            Error::new(
                ErrorKind::Unexpected,
                "invalid totalBytesRewritten in rewrite response",
            )
            .with_context("totalBytesRewritten", &result.total_bytes_rewritten)
            .set_source(err)
        })?;

        let progress = total
            .checked_sub(self.total_bytes_rewritten)
            .ok_or_else(|| {
                Error::new(
                    ErrorKind::Unexpected,
                    "rewrite response totalBytesRewritten moved backwards",
                )
                .with_context("previous", self.total_bytes_rewritten)
                .with_context("current", total)
            })?;

        self.total_bytes_rewritten = total;

        if result.done {
            self.completed = true;
            self.rewrite_token = None;
            self.metadata = Some(result.into_metadata(&self.to)?);
        } else {
            let token = result.rewrite_token.ok_or_else(|| {
                Error::new(
                    ErrorKind::Unexpected,
                    "rewrite response is not done but rewriteToken is missing",
                )
            })?;
            self.rewrite_token = Some(token);
        }

        if progress == 0 {
            return if self.completed {
                Ok(None)
            } else {
                Err(
                    Error::new(ErrorKind::Unexpected, "rewrite response made no progress")
                        .set_temporary(),
                )
            };
        }

        let progress = progress
            .try_into()
            .map_err(|_| Error::new(ErrorKind::Unexpected, "rewrite progress exceeds usize"))?;
        Ok(Some(progress))
    }

    async fn close(&mut self) -> Result<Metadata> {
        while !self.completed {
            self.next().await?;
        }

        Ok(self.metadata.clone().unwrap_or_default())
    }

    async fn abort(&mut self) -> Result<()> {
        self.rewrite_token = None;
        self.completed = true;
        self.metadata = None;
        Ok(())
    }
}

File: bindings/go/copier.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package opendal

import (
	"context"
	"runtime"
	"unsafe"

	"github.com/jupiterrider/ffi"
)

// Copier drives a long-running copy operation from a source to a destination.
//
// Unlike Copy, which performs the whole copy in a single call, a Copier lets the
// caller advance the copy step by step via Next, which is useful for reporting
// progress on large copies. The copy is complete once Next reports that no more
// progress is available. Always call Close to release the resources held by the
// Copier.
//
// # Example
//
//	func exampleCopier(op *opendal.Operator) {
//		copier, err := op.Copier("path/from/file", "path/to/file")
//		if err != nil {
//			log.Fatal(err)
//		}
//		defer copier.Close()
//		for {
//			n, ok, err := copier.Next()
//			if err != nil {
//				log.Fatal(err)
//			}
//			if !ok {
//				break
//			}
//			log.Printf("copied %d bytes", n)
//		}
//	}
type Copier struct {
	inner *opendalCopier
	ctx   context.Context
}

// Copier creates a Copier to copy a file from src to dest.
//
// Copier is a wrapper around the C-binding function `opendal_operator_copier`.
// When options are provided, it uses `opendal_operator_copier_with`.
//
// # Parameters
//
//   - src: The source file path.
//   - dest: The destination file path.
//   - opts: Optional copy options (shared with Copy).
//
// # Returns
//
//   - *Copier: A Copier used to drive the copy, or an error if the operation fails.
//
// # Behavior
//
//   - Both src and dest must be file paths, not directories.
//   - The returned Copier must be released with Close.
func (op *Operator) Copier(src, dest string, opts ...WithCopyFn) (*Copier, error) {
	if len(opts) == 0 {
		inner, err := ffiOperatorCopier.symbol(op.ctx)(op.inner, src, dest)
		if err != nil {
			return nil, err
		}
		return &Copier{inner: inner, ctx: op.ctx}, nil
	}

	o := parseCopyOptions(opts...)
	cOpts, keepAlive, err := newOpendalCopyOptions(op.ctx, o)
	if err != nil {
		return nil, err
	}
	defer ffiCopyOptionsFree.symbol(op.ctx)(cOpts)
	inner, err := ffiOperatorCopierWith.symbol(op.ctx)(op.inner, src, dest, cOpts)
	// cOpts holds raw pointers into the Go buffers tracked by keepAlive.
	// Keep them reachable until the native call above has returned.
	runtime.KeepAlive(keepAlive)
	if err != nil {
		return nil, err
	}
	return &Copier{inner: inner, ctx: op.ctx}, nil
}

// Next drives the copy forward by one step.
//
// It returns the number of bytes copied in this step. ok is false when the copy
// has completed; once that happens, further calls keep returning (0, false, nil).
// A non-nil error reports a failed copy. Calling Next on a closed Copier also
// returns (0, false, nil).
func (c *Copier) Next() (size uint, ok bool, err error) {
	if c.inner == nil {
		return 0, false, nil
	}
	return ffiCopierNext.symbol(c.ctx)(c.inner)
}

// Abort aborts the pending copy operation.
//
// After Abort, the Copier should still be released with Close. Calling Abort
// on a closed Copier is a no-op.
func (c *Copier) Abort() error {
	if c.inner == nil {
		return nil
	}
	return ffiCopierAbort.symbol(c.ctx)(c.inner)
}

// Close releases the resources held by the Copier. It must be called once the
// Copier is no longer needed. Close is safe to call more than once; subsequent
// calls are no-ops.
func (c *Copier) Close() error {
	if c.inner == nil {
		return nil
	}
	ffiCopierFree.symbol(c.ctx)(c.inner)
	c.inner = nil
	return nil
}

var ffiOperatorCopier = newFFI(ffiOpts{
	sym:    "opendal_operator_copier",
	rType:  &typeResultOperatorCopier,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(op *opendalOperator, src, dest string) (*opendalCopier, error) {
	return func(op *opendalOperator, src, dest string) (*opendalCopier, error) {
		byteSrc, err := BytePtrFromString(src)
		if err != nil {
			return nil, err
		}
		byteDest, err := BytePtrFromString(dest)
		if err != nil {
			return nil, err
		}
		var result resultOperatorCopier
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&op),
			unsafe.Pointer(&byteSrc),
			unsafe.Pointer(&byteDest),
		)
		if result.error != nil {
			return nil, parseError(ctx, result.error)
		}
		return result.copier, nil
	}
})

var ffiOperatorCopierWith = newFFI(ffiOpts{
	sym:    "opendal_operator_copier_with",
	rType:  &typeResultOperatorCopier,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer, &ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(op *opendalOperator, src, dest string, opts *opendalCopyOptions) (*opendalCopier, error) {
	return func(op *opendalOperator, src, dest string, opts *opendalCopyOptions) (*opendalCopier, error) {
		byteSrc, err := BytePtrFromString(src)
		if err != nil {
			return nil, err
		}
		byteDest, err := BytePtrFromString(dest)
		if err != nil {
			return nil, err
		}
		var result resultOperatorCopier
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&op),
			unsafe.Pointer(&byteSrc),
			unsafe.Pointer(&byteDest),
			unsafe.Pointer(&opts),
		)
		if result.error != nil {
			return nil, parseError(ctx, result.error)
		}
		return result.copier, nil
	}
})

var ffiCopierNext = newFFI(ffiOpts{
	sym:    "opendal_copier_next",
	rType:  &typeResultCopierNext,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(c *opendalCopier) (uint, bool, error) {
	return func(c *opendalCopier) (uint, bool, error) {
		var result resultCopierNext
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&c),
		)
		if result.error != nil {
			return 0, false, parseError(ctx, result.error)
		}
		return result.size, result.has_next == 1, nil
	}
})

var ffiCopierAbort = newFFI(ffiOpts{
	sym:    "opendal_copier_abort",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(c *opendalCopier) error {
	return func(c *opendalCopier) error {
		var e *opendalError
		ffiCall(
			unsafe.Pointer(&e),
			unsafe.Pointer(&c),
		)
		return parseError(ctx, e)
	}
})

var ffiCopierFree = newFFI(ffiOpts{
	sym:    "opendal_copier_free",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(c *opendalCopier) {
	return func(c *opendalCopier) {
		ffiCall(
			nil,
			unsafe.Pointer(&c),
		)
	}
})
